server/pulp/plugins/util/publish_step.py

```diff
@@ -562,6 +563,77 @@ def process_lifecycle(self):
         return self._build_final_report()
 
 
+class UnitModelPluginStep(PluginStep):
+    """
+    Useful for performing actions on units that are defined as mongoengine models.
+    """
+    def __init__(self, step_type, model_classes, repo_content_unit_q=None, repo=None, conduit=None,
+                 config=None, working_dir=None, plugin_type=None, **kwargs):
+        """
+        :param step_type: The id of the step this processes
+        :type  step_type: str
+        :param model_classes:   list of ContentUnit subclasses that should be queried
+        :type  model_classes:   list
+        :param repo_content_unit_q: optional Q object that will be applied to the queries performed
+                                    against each ContentUnit class that gets passed in
+        :type  repo_content_unit_q: mongoengine.Q
+        :param repo: The repo being worked on
+        :type  repo: pulp.plugins.model.Repository
+        :param conduit: The conduit for the repo
+        :type  conduit: a conduit from pulp.plugins.conduits
+        :param config: The configuration
+        :type  config: PluginCallConfiguration
+        :param working_dir: The temp directory this step should use for processing
+        :type  working_dir: str
+        :param plugin_type: The type of the plugin
+        :type  plugin_type: str
+        """
+        super(UnitModelPluginStep, self).__init__(step_type, repo, conduit, config, working_dir,
+                                                  plugin_type, **kwargs)
+
+        self.model_classes = model_classes
+        self._repo_content_unit_q = repo_content_unit_q
+
+        # the corresponding publicly-accessible values get cached here
+        self._unit_querysets = None
+        self._total = None
+
+    def get_iterator(self):
+        """
+        This method returns a generator to loop over ContentUnits. The items created by this
+        generator will be iterated over by the process_main method.
+
+        :return: a generator of ContentUnit objects
+        :rtype:  generator
+        """
+        return itertools.chain(*self.unit_querysets)
+
+    @property
+    def unit_querysets(self):
+        """
+        :return: list of QuerySet objects that correspond to ContentUnit searches for the classes
+                 passed in to this step.
+        :rtype:  list
+        """
+        if self._unit_querysets is None:
+            self._unit_querysets = []
+            for model_class in self.model_classes:
+                queries = repo_controller.get_unit_model_querysets(self.get_repo().id,
+                                                                   model_class,
+                                                                   self._repo_content_unit_q)
+                self._unit_querysets.extend(queries)
```

Comment:
I see now that `extend` does evaluate the iterable, but the iterable itself is a generator of `Queryset` objects. Thus your caching of them works with the mongoengine queryset caching and not against it. I consider this matter resolved.